### DISCUSSION TITLE
Alembic version out of sync between alembic vs db

### DISCUSSION THREAD

**Discussion by kingkong-cmd** [sqlalchemy/alembic#1094] [Usage Questions]
I use flask-migrate, and when I try to push (`flask db upgrade`) my latest migration file (`0db152ce17b6`) to the DB I get the below error:

`ERROR [flask_migrate] Error: Online migration expected to match one row when updating '3fba6dfbb433' to '0db152ce17b6' in 'alembic_version'; 0 found`

When I check my DB, the alembic version is `3fba6dfbb433`, but when I run `flask db heads` and `flask db history` i get the below results. I am not sure what the above error means, and also what it means that the head is different from the alembic version? Can anyone please help point me in the right direction to fix this?

```
flask db heads
0db152ce17b6 (head)
```
```
flask db history
3fba6dfbb433 -> 0db152ce17b6 (head), empty message
b1cb58b9ca83 -> 3fba6dfbb433, empty message
4b27b6043409 -> b1cb58b9ca83, empty message
88f37cb5274b -> 4b27b6043409, empty message
```

**Comment by zzzeek** [comment]
turn on SQLAlchemy "debug" logging, if using alembic.ini this looks like:

```py
# Logging configuration
[loggers]
keys = root,sqlalchemy,alembic

[handlers]
keys = console

[formatters]
keys = generic

[logger_root]
level = WARN
handlers = console
qualname =

[logger_sqlalchemy]
level = DEBUG
handlers =
qualname = sqlalchemy.engine

```

then we would want to see the full debug SQL output for the above operation as well as a complete stack trace.

other things:

1. this is just one specific database?  do these errors occur on a clean database re-running new migrations?
2. what database / driver is this?
3. has this database been working "fine" and something happened?   what precipitated it suddenly not working?
4. is this a devleopment DB you can just rebuild?

**Reply by kingkong-cmd** [reply]
> this is just one specific database? do these errors occur on a clean database re-running new migrations?

Yes, but haven't tried on a clean one.
>what database / driver is this?

Postgres running locally on my windows machine. Using SQLalchemy as ORM.
>has this database been working "fine" and something happened? what precipitated it suddenly not working?

Yes, it has been working fine. I am not sure exactly how and when this happened. I have pushed the migration to my Prod DB a few days ago, which works fine, but now, when I went back to using the Dev DB after some days, this problem suddenly turned up.
>is this a devleopment DB you can just rebuild?

Yes, potentially, but it would take a lot of work (I think) to populate it with the existing data, so hoping I dont have to.

See the SQLalchemy DEBUG log below. I don't quite see anything here, but maybe you do?

```
INFO  [sqlalchemy.engine.Engine] select version()
INFO  [sqlalchemy.engine.Engine] [raw sql] {}
DEBUG [sqlalchemy.engine.Engine] Col ('version',)
DEBUG [sqlalchemy.engine.Engine] Row ('PostgreSQL 12.1, compiled by Visual C++ build 1914, 64-bit',)
INFO  [sqlalchemy.engine.Engine] select current_schema()
INFO  [sqlalchemy.engine.Engine] [raw sql] {}
DEBUG [sqlalchemy.engine.Engine] Col ('current_schema',)
DEBUG [sqlalchemy.engine.Engine] Row ('public',)
INFO  [sqlalchemy.engine.Engine] show standard_conforming_strings
INFO  [sqlalchemy.engine.Engine] [raw sql] {}
DEBUG [sqlalchemy.engine.Engine] Col ('standard_conforming_strings',)
DEBUG [sqlalchemy.engine.Engine] Row ('on',)
INFO  [alembic.runtime.migration] Context impl PostgresqlImpl.
INFO  [alembic.runtime.migration] Will assume transactional DDL.
INFO  [sqlalchemy.engine.Engine] BEGIN (implicit)
INFO  [sqlalchemy.engine.Engine] select relname from pg_class c join pg_namespace n on n.oid=c.relnamespace where pg_catalog.pg_table_is_visible(c.oid) and relname=%(name)s
INFO  [sqlalchemy.engine.Engine] [generated in 0.00100s] {'name': 'alembic_version'}
DEBUG [sqlalchemy.engine.Engine] Col ('relname',)
DEBUG [sqlalchemy.engine.Engine] Row ('alembic_version',)
INFO  [sqlalchemy.engine.Engine] SELECT alembic_version.version_num
FROM alembic_version
INFO  [sqlalchemy.engine.Engine] [generated in 0.00098s] {}
DEBUG [sqlalchemy.engine.Engine] Col ('version_num',)
DEBUG [sqlalchemy.engine.Engine] Row ('3fba6dfbb433',)
INFO  [alembic.runtime.migration] Running upgrade 3fba6dfbb433 -> 0db152ce17b6, empty message
INFO  [sqlalchemy.engine.Engine] ALTER TABLE stock_level ADD COLUMN diff_from_previous_physical INTEGER
INFO  [sqlalchemy.engine.Engine] [no key 0.00073s] {}
INFO  [sqlalchemy.engine.Engine] ALTER TABLE stock_level ADD COLUMN quantity_physical INTEGER
INFO  [sqlalchemy.engine.Engine] [no key 0.00078s] {}
INFO  [sqlalchemy.engine.Engine] UPDATE alembic_version SET version_num='0db152ce17b6' WHERE alembic_version.version_num = 
'3fba6dfbb433'
INFO  [sqlalchemy.engine.Engine] [generated in 0.00110s] {}
INFO  [sqlalchemy.engine.Engine] ROLLBACK
ERROR [flask_migrate] Error: Online migration expected to match one row when updating '3fba6dfbb433' to '0db152ce17b6' in 'alembic_version'; 0 found
```

**Reply by CaselIT** [reply]
This is quite strange. The migration is there as shown by the select output, but then the update does not find it.

Can you try re-installing sqlalchemy and the dbapi driver you are using (I'm guessing psycopg2 since the db is postgresql) to rule out problems there?

**Reply by kingkong-cmd** [reply]
Did an update of both, but still getting the same log output...

```
SQLAlchemy==1.4.41
psycopg2==2.9.4
```

**Reply by kingkong-cmd** [reply]
In the mean time I managed to make a dump of my production db and restore it as a new dev db, so while I did not fix the underlying problem I managed to circumvent it.

**Reply by zzzeek** [reply]
can you show us the output of "SELECT * FROM alembic_version" ?

**Reply by kingkong-cmd** [reply]
> can you show us the output of "SELECT * FROM alembic_version" ?

It's `3fba6dfbb433`

**Reply by zzzeek** [reply]
OK how about this:

```sql
test=> BEGIN;
BEGIN
test=*> UPDATE alembic_version SET version_num='0db152ce17b6' WHERE alembic_version.version_num = 
'3fba6dfbb433';
UPDATE 1
test=*> SELECT * FROM alembic_version;
 version_num  
--------------
 0db152ce17b6
(1 row)

test=*> ROLLBACK;
ROLLBACK
```

see that "UPDATE 1"?  that "1" should say "1", not "0"

**Reply by kingkong-cmd** [reply]
This gives me the below. Not sure if the transaction in progress is just me messing up the SQL.

```
WARNING:  there is already a transaction in progress
UPDATE 0
```

**Reply by zzzeek** [reply]
hah.   do this:

```
select pid, usename, state, query from pg_stat_activity;
```

look for other transactions that are hung opened, looking at the same table.    or just restart your database.
